src/Features/CSharp/Portable/UseIsNullCheck/CSharpUseIsNullCheckForCastAndEqualityOperatorDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,81 @@
+    {
+        public CSharpUseIsNullCheckForCastAndEqualityOperatorDiagnosticAnalyzer()
+            : base(IDEDiagnosticIds.UseIsNullCheckForCastAndEqualityOperatorDiagnosticId,
+                   CSharpFeaturesResources.Use_is_null_check,
+                   new LocalizableResourceString(nameof(FeaturesResources.Null_check_can_be_simplified), FeaturesResources.ResourceManager, typeof(FeaturesResources)))
+        {
+        }
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory()
+            => DiagnosticAnalyzerCategory.SemanticSpanAnalysis;
+
+        public override bool OpenFileOnly(Workspace workspace)
+            => false;
+
+        protected override void InitializeWorker(AnalysisContext context)
+            => context.RegisterSyntaxNodeAction(AnalyzeSyntax, SyntaxKind.EqualsExpression, SyntaxKind.NotEqualsExpression);
+
+        private void AnalyzeSyntax(SyntaxNodeAnalysisContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+
+            var semanticModel = context.SemanticModel;
+            var syntaxTree = semanticModel.SyntaxTree;
+
+            if (((CSharpParseOptions)syntaxTree.Options).LanguageVersion < LanguageVersion.CSharp7)
+            {
+                return;
+            }
+
+            var optionSet = context.Options.GetDocumentOptionSetAsync(syntaxTree, cancellationToken).GetAwaiter().GetResult();
+            if (optionSet == null)
+            {
+                return;
+            }
+
+            var option = optionSet.GetOption(CodeStyleOptions.PreferIsNullCheckOverCastAndEqualityOperator, semanticModel.Language);
+            if (!option.Value)
+            {
+                return;
+            }
+
+            var binaryExpression = (BinaryExpressionSyntax)context.Node;
+
+            if (!IsObjectCastAndReferenceEquals(semanticModel, binaryExpression.Left, binaryExpression.Right) &&
+                !IsObjectCastAndReferenceEquals(semanticModel, binaryExpression.Right, binaryExpression.Left))
+            {
+                return;
+            }
+
+            var severity = option.Notification.Value;
+            context.ReportDiagnostic(
+                Diagnostic.Create(
+                    GetDescriptorWithSeverity(severity), binaryExpression.GetLocation()));
+        }
+
+        private bool IsObjectCastAndReferenceEquals(
+            SemanticModel semanticModel, ExpressionSyntax left, ExpressionSyntax right)
+        {
+            if (left is CastExpressionSyntax castExpression &&
+                right.IsKind(SyntaxKind.NullLiteralExpression))
+            {
+                // make sure it's a cast to object, and that the thing we're casting actually has a type.
+                return semanticModel.GetTypeInfo(castExpression.Type).Type?.SpecialType == SpecialType.System_Object &&
+                       semanticModel.GetTypeInfo(castExpression.Expression).Type != null;
+            }
+
+            return false;
+        }
```

Comment:
please make methods that can be static static, it makes it easier to reason about and know what to expect from the code when reviewing. otherwise i need to look if there's any instance state here that i'm missing

Reply:
done.